File: Programmers/20231216/포화 이진트리.js
function solution(numbers) {
    const answer = []
    function checkBTree(b_str,start,end) {
        const mid = Math.floor((start+end)/2)
        const left_c = Math.floor((start + mid-1)/2);
        const right_c = Math.floor((mid+1+ end)/2);

        if (start == end) {
            return true;
        }

        if(b_str[mid] == '0' && ((b_str[left_c] == '1')|| (b_str[right_c] == '1'))){
            return false;
        }

        if(!checkBTree(b_str, start, mid -1)) return false;
        if(!checkBTree(b_str, mid+1, end)) return false;
        return true; 
    }
    for (let number of numbers) {
        let bi_num = number.toString(2);
        let n = bi_num.length;
        // 0101010
        let m = n.toString(2).length //이진법으로 바기
        let bi_tree = '0'.repeat(2**m-1 - n)
        bi_tree = bi_tree +'' + bi_num;
        
        if(checkBTree(bi_tree,0,bi_tree.length-1)) {
            answer.push(1);
        }
        else{
            answer.push(0);
        }


    }
    return answer
}

console.log(solution([7,42,5]))

File: Programmers/20231216/포화 이진트리.py
import math

def child_parent(num,parent) :
    if num == '':
        return True
    mid = len(num)//2
    me = num[mid]
    if me == '1' and parent == '0':
        return False
    else:
        return child_parent(num[:mid],me) and child

def check(sum):
    if num == 1:
        return 1
    num = bin(num)[2:]
    digit = 2 ** (int(math.log(len(num),2)) + 1) - 1
    num = "0" * (digit - len(num)) + num

    if num[len(num)//2] == '1' and child_parent(num,True):
        return 1
    else:
        return 0

def solution(numbers):
    answer =[]
    for num in numbers:
        answer.append(check(sum))
    return answer